spec/formatting.md

```diff
@@ -2,13 +2,16 @@
 
 ## Introduction
 
-This document defines the behaviour of a MessageFormat 2.0 implementation
+This document defines the behavior of a MessageFormat 2.0 implementation
 when formatting a message for display in a user interface, or for some later processing.
 
-To start, we presume that a _message_ has either been parsed from its syntax
-or created from a data model description.
-If this construction has encountered any _Syntax Errors_ or _Data Model Errors_,
-an appropriate error MUST be emitted and a _fallback value_ MAY be used as the formatting result.
+To start, parse the _message_ from its syntax or create it from a data model description.
+
+If the message being formatted has any _Syntax Errors_ or _Data Model Errors_,
+an appropriate error MUST be emitted and
```

Comment:
This seems like it's requiring eager evaluation. It's reasonable for a lazy implementation to have to report syntax errors even in expressions that are never used. But when it comes to data model errors, I'm not sure. For example, duplicate option name errors: do we want to require an implementation to error out on:

```
.local $foo = {horse :func one=1 two=2 one=1}
{{I don't use any variables!}}
```

even though `$foo` is never referenced? Maybe this is a weird message and should always be an error. I don't have a strong opinion, I just want to make sure it's considered.

Reply:
One of the things I did in my mock is to discard variables that were never used. And since they were never used, I never even tried to check them further. I don't think we need to force errors to be omitted in those cases, although it could be useful to have a 'strict' mode that emitted warnings, etc.

BTW — and this is not _in any way_ suggesting a change in v45 — I have a somewhat different breakdown of errors, based information available at a given point. Categories 1-4 are all errors that can be handled through static analysis, before any runtime use.

1. Errors that can be detected syntactically, without reference to any function registry. 
  * That includes any BNF breakages
  * But other failures that don't require building a model or having any access to information about functions and their options:  a mismatch between number of keys in a variant and the number of selectors in the .match, no fallback variant, duplicate declarations or option names, ...
2. Errors that can be detected with reference to reference to the registry, but without having any input parameters
  * This includes illegal options, illegal option values, some illegal literals (that the function can't handle), .match with functions that are not selector functions, placeholders with functions that aren't formatting functions, etc.
3. Errors that require calling functions, but need no access to input parameters
  * Literals that can't be handled, but regex in the registry is not powerful enough to limit
  * Mismatches like .input {$person :personName ..} .local $date {$person :datetime ...), if the registry is not detailed enough to handle.
4. Errors that can be detected at call sites (in strongly-typed programming languages, depending on APIs)
   *  There are not enough input parameters, or the types are not accepted by the functions that will be called on them.
5. Runtime errors, that cannot be assessed except with particular input parameters
  * A function can't handle a particular input datatype, etc.

Reply:
@catamorphism  notes:

> This seems like it's requiring eager evaluation.

We have a very strong statement that we never require eager evaluation. As you note, syntax errors are not what is in question here. They are always errors and they have nothing to do with expression evaluation.

Of the data model errors, most are not related to eager vs. lazy evaluation--except duplicate option name and maybe missing selector annotation. That seems like something that should not be included in the MUST I suggest above. The others are structural errors with the message that have to be evaluated in order for pattern selection to run. This is in part why the text here was originally in the pattern selection section. 

Probably we should split syntax from data model error and push the DM errors back into pattern selection where they started. (Expression evaluation already has the one for duplicate option names)

@macchiati 

You're welcome to discard unused variables--including never evaluating the expressions assigned to them--in your implementation (we say this explicitly). However, of the data model errors (which are not related to anything in the function registry), as noted, only the duplication of options is really "looking inside" the expression. The missing selector annotation has to be checked because it is on the `.match` statement. Any message with a `.match` has to evaluate the selectors.

I have more thoughts, but not time tonight to delve into them.